README.md

```diff
@@ -122,15 +122,15 @@ A `MiddlewareFunc` preserves the default net/http middleware pattern.
 You can create new middleware functions and pass them to Service to be chained on all routes in the default Http Component.
 
 ```go
-type MiddlewareFunc func(next http.HandlerFunc) http.HandlerFunc
+type MiddlewareFunc func(next http.Handler) http.Handler
```

Comment:
the most common type of a middleware follows the previous signature:
```go
type MiddlewareFunc func(next http.HandlerFunc) http.HandlerFunc
```
all existing middleware uses this type

Reply:
> the most common type of a middleware follows the previous signature ...

`HandlerFunc` and `Handler` are _interchangeable_.
As we discussed previously it is just a concrete implementation of `Handler`
This way it is very easy to wrap `httprouter.Router` that is an other implementation of `Handler`.
So by using the interface and not a concrete implementation we lose a bit of simplicity but we can use the same chaining mechanism for routes (handlerfunc) as well in our further development.

> all existing middleware uses this type

We will change the existing middlewares to follow the `MiddlewareFunc` signature (which they don't now) and they will be "chained" the same way.

Reply:
They might be interchangeable but are not the same.
Handler is an interface(you need to have a something that implements the ServeHTTP method):
```go
type Handler interface {
        ServeHTTP(ResponseWriter, *Request)
}
```
where as HandlerFunc is just a function type:
```go
type HandlerFunc func(ResponseWriter, *Request)
```
and the two look like this when implemented:
HandlerFunc
```go
func middle1(rw http.ResponseWriter, r *http.Request) {
}
```
Handler
```go
func middle2(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		h.ServeHTTP(w, r)
	})
}
```
Definetely HandlerFunc is a much better choice.

Reply:
> where as HandlerFunc is just a function type:
` HandlerFunc` implements `Handler`, it is a receiver of the ServeHTTP method
I agree that we loose simplicity (just a bit - I wrote that in my comment) but we gain a more elegant and generic solution.

Reply:
Perfect. Let's go forward with that.